File: src/lib/toml/validate.ts
/**
 * Validation logic for vmsan.toml configuration files.
 *
 * Checks: syntax, missing dependencies, circular dependencies,
 * invalid ports, unknown runtimes, missing required fields,
 * and duplicate service names.
 */

import { parse, TomlError } from "smol-toml";
import type { VmsanToml, ServiceConfig } from "./types.ts";
import { findClosestMatch as findClosest } from "./suggest.ts";
import { VALID_RUNTIMES } from "../../commands/create/types.ts";

export interface ValidationError {
  line?: number;
  field: string;
  message: string;
  suggestion?: string;
}

/**
 * Parse raw TOML text and return the config or syntax errors.
 */
export function parseTomlSafe(tomlText: string): {
  config: VmsanToml | null;
  errors: ValidationError[];
} {
  try {
    const config = parse(tomlText) as unknown as VmsanToml;
    return { config, errors: [] };
  } catch (error) {
    if (error instanceof TomlError) {
      return {
        config: null,
        errors: [
          {
            line: error.line,
            field: "syntax",
            message: error.message,
          },
        ],
      };
    }
    return {
      config: null,
      errors: [
        {
          field: "syntax",
          message: error instanceof Error ? error.message : String(error),
        },
      ],
    };
  }
}

/**
 * Validate a parsed VmsanToml config. Returns an array of validation errors.
 * An empty array means the config is valid.
 */
export function validateToml(config: VmsanToml): ValidationError[] {
  const errors: ValidationError[] = [];

  // Collect all service + accessory names
  const allNames = new Set<string>();
  const serviceNames = new Set(Object.keys(config.services ?? {}));
  const accessoryNames = new Set(Object.keys(config.accessories ?? {}));

  for (const name of serviceNames) {
    allNames.add(name);
  }
  for (const name of accessoryNames) {
    allNames.add(name);
  }

  // Check for duplicate names across services and accessories
  for (const name of serviceNames) {
    if (accessoryNames.has(name)) {
      errors.push({
        field: `services.${name}`,
        message: `Duplicate name "${name}" — used in both services and accessories`,
      });
    }
  }

  // Validate services
  if (config.services) {
    for (const [name, service] of Object.entries(config.services)) {
      validateService(name, service, allNames, errors);
    }

    // Check circular dependencies
    const circularErrors = detectCircularDependencies(config.services);
    errors.push(...circularErrors);
  }

  return errors;
}

function validateService(
  name: string,
  service: ServiceConfig,
  allNames: Set<string>,
  errors: ValidationError[],
): void {
  // Missing start command
  if (!service.start) {
    errors.push({
      field: `services.${name}.start`,
      message: `Service "${name}" is missing a "start" command`,
    });
  }

  // Unknown runtime
  if (
    service.runtime &&
    !VALID_RUNTIMES.includes(service.runtime as (typeof VALID_RUNTIMES)[number])
  ) {
    const suggestion = findClosestMatch(service.runtime, [...VALID_RUNTIMES]);
    errors.push({
      field: `services.${name}.runtime`,
      message: `Unknown runtime "${service.runtime}" in service "${name}"`,
      suggestion: suggestion
        ? `Did you mean "${suggestion}"?`
        : `Valid runtimes: ${VALID_RUNTIMES.join(", ")}`,
    });
  }

  // Missing dependencies
  if (service.depends_on) {
    for (const dep of service.depends_on) {
      if (!allNames.has(dep)) {
        const suggestion = findClosestMatch(dep, [...allNames]);
        errors.push({
          field: `services.${name}.depends_on`,
          message: `Service "${name}" references unknown service "${dep}" in depends_on`,
          suggestion: suggestion ? `Did you mean "${suggestion}"?` : undefined,
        });
      }
    }
  }

  // Invalid ports in connect_to
  if (service.connect_to) {
    for (const entry of service.connect_to) {
      const colonIndex = entry.lastIndexOf(":");
      if (colonIndex === -1) {
        errors.push({
          field: `services.${name}.connect_to`,
          message: `Invalid connect_to entry "${entry}" in service "${name}": expected format "service:port"`,
        });
        continue;
      }

      const targetService = entry.slice(0, colonIndex);
      const portStr = entry.slice(colonIndex + 1);
      const port = Number(portStr);

      if (!allNames.has(targetService)) {
        const suggestion = findClosestMatch(targetService, [...allNames]);
        errors.push({
          field: `services.${name}.connect_to`,
          message: `Service "${name}" references unknown service "${targetService}" in connect_to`,
          suggestion: suggestion ? `Did you mean "${suggestion}"?` : undefined,
        });
      }

      if (Number.isNaN(port) || port < 1 || port > 65535 || !Number.isInteger(port)) {
        errors.push({
          field: `services.${name}.connect_to`,
          message: `Invalid port "${portStr}" in connect_to entry "${entry}" for service "${name}"`,
          suggestion: "Port must be an integer between 1 and 65535",
        });
      }
    }
  }
}

/**
 * Detect circular dependencies in the service dependency graph.
 */
function detectCircularDependencies(services: Record<string, ServiceConfig>): ValidationError[] {
  const errors: ValidationError[] = [];
  const visited = new Set<string>();
  const inStack = new Set<string>();

  function dfs(name: string, path: string[]): boolean {
    if (inStack.has(name)) {
      const cycleStart = path.indexOf(name);
      const cycle = path.slice(cycleStart).concat(name);
      errors.push({
        field: "services",
        message: `Circular dependency detected: ${cycle.join(" -> ")}`,
      });
      return true;
    }

    if (visited.has(name)) return false;

    visited.add(name);
    inStack.add(name);

    const service = services[name];
    if (service?.depends_on) {
      for (const dep of service.depends_on) {
        if (services[dep]) {
          dfs(dep, [...path, name]);
        }
      }
    }

    inStack.delete(name);
    return false;
  }

  for (const name of Object.keys(services)) {
    if (!visited.has(name)) {
      dfs(name, []);
    }
  }

  return errors;
}

function findClosestMatch(input: string, candidates: string[]): string | null {
  return findClosest(input, candidates);
}

File: hostd/internal/config/validate.go
package config

import (
	"fmt"
	"net"
	"strconv"
	"strings"
)

// ValidationError represents a single validation issue.
type ValidationError struct {
	Line       int    `json:"line,omitempty"`
	Field      string `json:"field"`
	Message    string `json:"message"`
	Suggestion string `json:"suggestion,omitempty"`
}

// ParseTomlSafe parses TOML and returns both the config and any syntax errors.
func ParseTomlSafe(tomlText string) (*VmsanToml, []ValidationError) {
	cfg, err := ParseVmsanToml(tomlText)
	if err != nil {
		return nil, []ValidationError{{
			Field:   "syntax",
			Message: err.Error(),
		}}
	}
	return cfg, nil
}

// ValidateToml validates a parsed VmsanToml config and returns all errors found.
func ValidateToml(cfg *VmsanToml) []ValidationError {
	var errs []ValidationError
	services := NormalizeToml(cfg)

	for name, svc := range services {
		errs = append(errs, validateService(name, svc, services, cfg.Accessories)...)
	}
	for name, acc := range cfg.Accessories {
		errs = append(errs, validateAccessory(name, acc)...)
	}

	// Check for circular dependencies
	if cycles := detectCycles(services, cfg.Accessories); len(cycles) > 0 {
		for _, cycle := range cycles {
			errs = append(errs, ValidationError{
				Field:   "depends_on",
				Message: fmt.Sprintf("Circular dependency detected: %s", strings.Join(cycle, " -> ")),
			})
		}
	}

	// Check for duplicate names across services and accessories
	for name := range services {
		if _, ok := cfg.Accessories[name]; ok {
			errs = append(errs, ValidationError{
				Field:   fmt.Sprintf("services.%s", name),
				Message: fmt.Sprintf("Duplicate name %q used in both services and accessories", name),
			})
		}
	}

	return errs
}

func validateService(name string, svc ServiceConfig, allServices map[string]ServiceConfig, accessories map[string]AccessoryConfig) []ValidationError {
	var errs []ValidationError

	// Missing start command
	if svc.Start == "" {
		errs = append(errs, ValidationError{
			Field:      fmt.Sprintf("services.%s.start", name),
			Message:    fmt.Sprintf("Service %q is missing a start command", name),
			Suggestion: "Add start = \"your-start-command\" to the service",
		})
	}

	// Unknown runtime
	if svc.Runtime != "" && !isValidRuntime(svc.Runtime) {
		suggestion := FindClosestMatch(svc.Runtime, ValidRuntimes)
		sugMsg := fmt.Sprintf("Valid runtimes: %s", strings.Join(ValidRuntimes, ", "))
		if suggestion != "" {
			sugMsg = fmt.Sprintf("Did you mean %q?", suggestion)
		}
		errs = append(errs, ValidationError{
			Field:      fmt.Sprintf("services.%s.runtime", name),
			Message:    fmt.Sprintf("Unknown runtime %q in service %q", svc.Runtime, name),
			Suggestion: sugMsg,
		})
	}

	// Missing dependency references
	for _, dep := range svc.DependsOn {
		if _, ok := allServices[dep]; !ok {
			if _, ok := accessories[dep]; !ok {
				errs = append(errs, ValidationError{
					Field:   fmt.Sprintf("services.%s.depends_on", name),
					Message: fmt.Sprintf("Service %q depends on %q which is not defined", name, dep),
				})
			}
		}
	}

	// Invalid ports in connect_to
	for _, ct := range svc.ConnectTo {
		parts := strings.SplitN(ct, ":", 2)
		if len(parts) == 2 {
			port, err := strconv.Atoi(parts[1])
			if err != nil || port < 1 || port > 65535 {
				errs = append(errs, ValidationError{
					Field:      fmt.Sprintf("services.%s.connect_to", name),
					Message:    fmt.Sprintf("Invalid port in connect_to %q for service %q", ct, name),
					Suggestion: "Ports must be between 1 and 65535",
				})
			}
		}
	}

	// Invalid publish_ports
	for _, port := range svc.PublishPorts {
		if port < 1 || port > 65535 {
			errs = append(errs, ValidationError{
				Field:   fmt.Sprintf("services.%s.publish_ports", name),
				Message: fmt.Sprintf("Invalid port %d in service %q", port, name),
			})
		}
	}

	// Validate network_policy
	if svc.NetworkPolicy != "" && !isValidNetworkPolicy(svc.NetworkPolicy) {
		errs = append(errs, ValidationError{
			Field:      fmt.Sprintf("services.%s.network_policy", name),
			Message:    fmt.Sprintf("Unknown network_policy %q in service %q", svc.NetworkPolicy, name),
			Suggestion: fmt.Sprintf("Valid policies: %s", strings.Join(ValidNetworkPolicies, ", ")),
		})
	}

	// Validate allowed_domains requires custom policy
	if len(svc.AllowedDomains) > 0 && svc.NetworkPolicy != "" && svc.NetworkPolicy != "custom" {
		errs = append(errs, ValidationError{
			Field:      fmt.Sprintf("services.%s.allowed_domains", name),
			Message:    fmt.Sprintf("Service %q has allowed_domains but network_policy is %q", name, svc.NetworkPolicy),
			Suggestion: "Set network_policy = \"custom\" to use allowed_domains",
		})
	}

	// Validate CIDR-like values in allowed_domains
	for _, domain := range svc.AllowedDomains {
		if _, _, err := net.ParseCIDR(domain); err == nil {
			errs = append(errs, ValidationError{
				Field:      fmt.Sprintf("services.%s.allowed_domains", name),
				Message:    fmt.Sprintf("Value %q looks like a CIDR, not a domain", domain),
				Suggestion: "Use allowed_cidrs for IP ranges",
			})
		}
	}

	return errs
}

func validateAccessory(name string, acc AccessoryConfig) []ValidationError {
	var errs []ValidationError

	if acc.Type == "" {
		errs = append(errs, ValidationError{
			Field:   fmt.Sprintf("accessories.%s.type", name),
			Message: fmt.Sprintf("Accessory %q is missing required field 'type'", name),
		})
	} else if !isValidAccessoryType(acc.Type) {
		errs = append(errs, ValidationError{
			Field:      fmt.Sprintf("accessories.%s.type", name),
			Message:    fmt.Sprintf("Unknown accessory type %q", acc.Type),
			Suggestion: fmt.Sprintf("Valid types: %s", strings.Join(ValidAccessoryTypes, ", ")),
		})
	}

	return errs
}

// detectCycles uses DFS three-color marking to find circular dependencies.
func detectCycles(services map[string]ServiceConfig, accessories map[string]AccessoryConfig) [][]string {
	const (
		white = 0 // unvisited
		gray  = 1 // visiting
		black = 2 // done
	)

	color := make(map[string]int)
	parent := make(map[string]string)
	var cycles [][]string

	// Collect all node names
	for name := range services {
		color[name] = white
	}
	for name := range accessories {
		color[name] = white
	}

	var dfs func(node string)
	dfs = func(node string) {
		color[node] = gray

		var deps []string
		if svc, ok := services[node]; ok {
			deps = svc.DependsOn
		}

		for _, dep := range deps {
			if _, exists := color[dep]; !exists {
				continue // unknown dep, caught by other validation
			}
			switch color[dep] {
			case white:
				parent[dep] = node
				dfs(dep)
			case gray:
				// Build cycle path
				cycle := []string{dep}
				cur := node
				for cur != dep {
					cycle = append([]string{cur}, cycle...)
					cur = parent[cur]
				}
				cycle = append([]string{dep}, cycle...)
				cycles = append(cycles, cycle)
			}
		}

		color[node] = black
	}

	for name := range color {
		if color[name] == white {
			dfs(name)
		}
	}

	return cycles
}

func isValidRuntime(r string) bool {
	for _, v := range ValidRuntimes {
		if v == r {
			return true
		}
	}
	return false
}

func isValidNetworkPolicy(p string) bool {
	for _, v := range ValidNetworkPolicies {
		if v == p {
			return true
		}
	}
	return false
}

func isValidAccessoryType(t string) bool {
	for _, v := range ValidAccessoryTypes {
		if v == t {
			return true
		}
	}
	return false
}
